lisa/tools/hugepages.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import re
+from typing import Any, Set
+
+from lisa.executable import Tool
+from lisa.tools.echo import Echo
+from lisa.tools.free import Free
+from lisa.tools.ls import Ls
+from lisa.tools.lscpu import Lscpu
+from lisa.tools.mkfs import FileSystem
+from lisa.tools.mount import Mount
+from lisa.util import LisaException, SkippedException
+
+PATTERN_HUGEPAGE = re.compile(
+    r"^.*hugepages-(?P<hugepage_size_in_kb>\d+)kB.*",
+)
+
+
+class Hugepages(Tool):
+    @property
+    def command(self) -> str:
+        return ""
+
+    @property
+    def can_install(self) -> bool:
+        return False
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._hugepage_sizes: Set[str] = set()
+        self._hugepages_dir_path = "/sys/devices/system/node/node*/hugepages"
+
+    def get_hugepage_sizes_in_kb(self) -> Set[str]:
+        """
+        This function get the hugepage sizes in kB available on the system.
+        This is done by listing the hugepages directory
+        and extracting the size from the directory name.
+        e.g. [
+        "'/sys/devices/system/node/node0/hugepages/hugepages-16777216kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-2048kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-524288kB/'"
+        ]
+        """
+        ls = self.node.tools[Ls]
+        # if not ls.path_exists(self._hugepages_dir_path, sudo=True):
+        #     raise SkippedException(f"path: {self._hugepages_dir_path} does not exist")
+        hugepage_size_dir_names = ls.list_dir(self._hugepages_dir_path, sudo=True)
+        for hugepage_size_dir_name in hugepage_size_dir_names:
+            matched_hugepage = PATTERN_HUGEPAGE.match(hugepage_size_dir_name)
+            if not matched_hugepage:
+                raise LisaException(
+                    f"no hugepage size found in {hugepage_size_dir_name}"
+                )
+            self._hugepage_sizes.add(matched_hugepage.group("hugepage_size_in_kb"))
+        return self._hugepage_sizes
+
+    def _enable_hugepages(self, enable_gibibyte_hugepages: bool = False) -> None:
+        echo = self.node.tools[Echo]
+
+        meminfo = self.node.tools[Free]
+        nics_count = len(self.node.nics.get_nic_names())
+
+        numa_nodes = self.node.tools[Lscpu].get_numa_node_count()
+        request_pages_2mb = (nics_count - 1) * 1024 * numa_nodes
+        request_pages_1gb = (nics_count - 1) * numa_nodes
+        memfree_2mb = meminfo.get_free_memory_mb()
+        memfree_1mb = meminfo.get_free_memory_gb()
+
+        # request 2iGB memory per nic, 1 of 2MiB pages and 1 GiB page
+        # check there is enough memory on the device first.
+        # default to enough for one nic if not enough is available
+        # this should be fine for tests on smaller SKUs
+
+        if enable_gibibyte_hugepages:
```

Comment:
There HAS to be some easier way to do this. Don't feel obligated to rewrite this verbose nonsense I wrote exactly.

Like, attempting to enable and checking the return code of the write to sysfs? Or attempting to enable and checking if you got the new amount you asked for? I'm sure there's a manpage somewhere with advice on how to check whether the size you asked for was enabled properly.

Reply:
check if included in the tool